File: mathewDavidov/assignments/treesBFS/lc103/lc103.py
# Problem Statement #

# Given a binary tree, populate an array to represent its zigzag level order traversal. You should populate the values of all nodes of the first level from left to right, then right to left for the next level and keep alternating in the same manner for the following levels.

class TreeNode:
  def __init__(self, val):
    self.val = val
    self.left, self.right = None, None

def traverse(root):
  result = []
  # TODO: Write your code here
  return result


def main():
  root = TreeNode(12)
  root.left = TreeNode(7)
  root.right = TreeNode(1)
  root.left.left = TreeNode(9)
  root.right.left = TreeNode(10)
  root.right.right = TreeNode(5)
  root.right.left.left = TreeNode(20)
  root.right.left.right = TreeNode(17)
  print("Zigzag traversal: " + str(traverse(root)))


main()



# Solution
# -----
# def traverse(root):
# 	if not root: return []
# 	queue = collections.deque([root])
# 	res = []
# 	even_level = False
# 	while queue:
# 		n = len(queue)
# 		level = []
# 		for _ in range(n):
# 			node = queue.popleft()
# 			level.append(node.val)
# 			if node.left:
# 				queue.append(node.left)
# 			if node.right:
# 				queue.append(node.right)
# 		if even_level:
# 			res.append(level[::-1])
# 		else:
# 			res.append(level)
# 		even_level = not even_level
# 	return res

# -----

# Time complexity #
# The time complexity of the above algorithm is O(N), where ‘N’ is the total number of nodes in the tree. This is due to the fact that we traverse each node once.

# Space complexity #
# The space complexity of the above algorithm will be O(N) as we need to return a list containing the level order traversal. We will also need O(N) space for the queue. Since we can have a maximum of N/2 nodes at any level (this could happen only at the lowest level), therefore we will need O(N) space to store them in the queue.

File: mathewDavidov/assignments/treesBFS/lc103/lc103.java
/**
 * Problem Statement:
 * Given a binary tree, populate an array to represent its zigzag level order traversal. 
 * You should populate the values of all nodes of the first level from left to right, 
 * then right to left for the next level and keep alternating in the same manner for the following levels.
 * 
 * Example:
 * Input: root = [3,9,20,null,null,15,7]
 *       3
 *      9 20
 *       15 7
 * Output: [[3],[20,9],[15,7]]
 * 
 * Edge cases:
 * root null -> return empty list
 * 
 * initialize a list containing the level-by-level traversal
 * initialize a queue and add the root
 * while the queue is not empty:
 *   retrieve the current size of the queue in a variable
 *   initialize a list for the current level (linkedlist, since we will take advantage of adding to the front)
 *   loop from 0 to the size of the queue (in the above variable):
 *     poll the element in the front in the queue
 *     if its left child exists, add it to the queue
 *     if its right child exists, add it to the queue
 *     if the level-by-level list size is even, simply add the value to the current layer's list
 *     otherwise, add the value to the front of the current layer's list, which provides the alternating pattern
 *   add the current level list to the front of the level-by-level list
 * return the level-by-level list
 * 
 * Runtime:
 * Time: O(n) - note: adding to the front of a linkedlist is O(1)
 * Space: O(n)
 */
import java.util.*;

public class lc103 {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> levelByLevelList = new ArrayList<>();

        if (root == null) {
            return levelByLevelList;
        }

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int queueSize = queue.size();
            List<Integer> levelList = new LinkedList<>();

            for (int i = 0; i < queueSize; i++) {
                TreeNode current = queue.poll();

                if (current.left != null) {
                    queue.offer(current.left);
                }

                if (current.right != null) {
                    queue.offer(current.right);
                }

                if (levelByLevelList.size() % 2 == 0) {
                    levelList.add(current.val);
                } else {
                    levelList.add(0, current.val);
                }
            }

            levelByLevelList.add(levelList);
        }

        return levelByLevelList;
    }

    public static void main(String[] args) {

    }
}

// class TreeNode {
//     public int val = 0;
//     public TreeNode left;
//     public TreeNode right;
   
//     public TreeNode(int val) {
//        this.val = val;
//        this.left = this.right = null;
//     }
// }